src/Rings/oscar_singular.jl

```diff
@@ -0,0 +1,458 @@
+#
+# a real native Singular polynomial ring with Singular's native QQ as coefficients:
+#  singular_poly_ring(QQ[t]) => Singular.PolyRing{Singular.n_Q}
+#
+# Singular's native Fp(5):
+#  singular_coeff_ring(GF(5)) => Singular.N_ZpField
+#
+# Singular wrapper of the Oscar type QQPolyRingElem:
+#  singular_coeff_ring(QQ[t]) => Singular.N_Ring{QQPolyRingElem}
+#
+# even more wrappings of the immutable Oscar type FpFieldElem:
+#  singular_coeff_ring(GF(ZZRingElem(5))) => Singular.N_Field{Singular.FieldElemWrapper{FpField, FpFieldElem}}
+
+"""
+    iso_oscar_singular_coeff_ring(R::Ring) -> Map
+
+Return a ring isomorphism of `R` onto a native Singular ring
+"""
+iso_oscar_singular_coeff_ring
+
+"""
+    iso_oscar_singular_poly_ring(R::Ring, ...; kw...) -> Map
+
+Given a polynomial ring `R[x_1,...x_n]` return a ring isomorphism onto a native
+Singular ring `S[x_1,...,x_n]` with `R` isomorphic to `S`.
+"""
+iso_oscar_singular_poly_ring
+
+abstract type OscarSingularCoefficientRingMap{S, T} <: Map{S, T, Any, Any} end
+
+domain(f::OscarSingularCoefficientRingMap) = f.R
+
+codomain(f::OscarSingularCoefficientRingMap) = f.S
+
+# fallback
+function image(f::OscarSingularCoefficientRingMap, x)
+  parent(x) !== domain(f) && error("Element not in domain")
+  return codomain(f)(x)
+end
+
+(f::OscarSingularCoefficientRingMap)(x) = image(f, x)
+
+function preimage(f::OscarSingularCoefficientRingMap, x)
+  parent(x) !== codomain(f) && error("Element not in codomain")
+  return domain(f)(x)
+end
+
+# ZZ
+struct OscarSingularCoefficientRingMapZZ <: OscarSingularCoefficientRingMap{ZZRing, Singular.Integers} 
+  R::ZZRing # singleton, but keep for consistency
+  S::Singular.Integers
+end
+
+iso_oscar_singular_coeff_ring(R::ZZRing) = OscarSingularCoefficientRingMapZZ(R, Singular.Integers())
+
+# conversion is unambigous via parent object call
+
+# QQ
+struct OscarSingularCoefficientRingMapQQ <: OscarSingularCoefficientRingMap{QQField, Singular.Rationals}
+  R::QQField # singleton, but keep for consistency
+  S::Singular.Rationals
+end
+
+iso_oscar_singular_coeff_ring(R::QQField) = OscarSingularCoefficientRingMapQQ(R, Singular.Rationals())
+
+# prime field, small characteristic
+struct OscarSingularCoefficientRingMapfpField <: OscarSingularCoefficientRingMap{fpField, Singular.N_ZpField}
+  R::fpField
+  S::Singular.N_ZpField
+end
+
+iso_oscar_singular_coeff_ring(R::fpField) = OscarSingularCoefficientRingMapfpField(R, Singular.Fp(Int(characteristic(R))))
+
+# ZZ/nZZ, n small and big
+struct OscarSingularCoefficientRingMapZZMod{U} <: OscarSingularCoefficientRingMap{U, Singular.N_ZpField}
+  R::U
+  S::Singular.N_ZnRing
+end
+
+iso_oscar_singular_coeff_ring(R::U) where {U <: Union{zzModRing, ZZModRing}} = OscarSingularCoefficientRingMapZZMod{U}(R, Singular.residue_ring(Singular.Integers(), BigInt(modulus(R)))[1]
+)
+
+# QQ(a)
+struct OscarSingularCoefficientRingMapAbsSimpleNumField <: OscarSingularCoefficientRingMap{AbsSimpleNumField, Singular.N_ZpField}
```

Comment:
Same question as above: should this perhaps be
```suggestion
struct OscarSingularCoefficientRingMapAbsSimpleNumField <: OscarSingularCoefficientRingMap{AbsSimpleNumField, Singular.N_AlgExtField}
```
If not then perhaps a comment explaining it would be helpful